pkg/ccl/changefeedccl/testfeed_test.go

```diff
@@ -1041,52 +1058,141 @@ func extractKeyFromJSONValue(isBare bool, wrapped []byte) (key []byte, value []b
 	return key, value, nil
 }
 
-// Next implements the TestFeed interface.
-func (c *cloudFeed) Next() (*cdctest.TestFeedMessage, error) {
+// This function reads the parquet file and converts each row to its JSON
+// equivalent and appends it to the cloudfeed's row object.
+func (c *cloudFeed) appendParquetTestFeedMessages(path string, topic string) error {
+
+	f, err := os.Open(path)
+	if err != nil {
+		return err
+	}
+	defer f.Close()
+
+	fr, err := goparquet.NewFileReader(f)
+	if err != nil {
+		return err
+	}
+
+	keyCols, ok := fr.MetaData()["pkeys"]
+	if !ok {
+		return err
+	}
+	keyColsSet := make(map[string]struct{})
+	for _, key := range strings.Split(keyCols, ",") {
+		keyColsSet[key] = struct{}{}
+	}
+
 	for {
-		if len(c.rows) > 0 {
-			e := c.rows[0]
-			c.rows = c.rows[1:]
-			m := &cdctest.TestFeedMessage{
-				Topic:    e.topic,
-				Value:    e.value,
-				Resolved: e.payload,
-			}
 
-			// The other TestFeed impls check both key and value here, but cloudFeeds
-			// don't have keys.
-			if len(m.Value) > 0 {
-				details, err := c.Details()
-				if err != nil {
-					return nil, err
-				}
+		row, err := fr.NextRow()
+		if err == io.EOF {
+			break
+		}
 
-				switch v := changefeedbase.FormatType(details.Opts[changefeedbase.OptFormat]); v {
-				case ``, changefeedbase.OptFormatJSON:
-					// Cloud storage sinks default the `WITH key_in_value` option so that
-					// the key is recoverable. Extract it out of the value (also removing it
-					// so the output matches the other sinks). Note that this assumes the
-					// format is json, this will have to be fixed once we add format=avro
-					// support to cloud storage.
-					//
-					// TODO(dan): Leave the key in the value if the TestFeed user
-					// specifically requested it.
-					if m.Key, m.Value, err = extractKeyFromJSONValue(c.isBare, m.Value); err != nil {
-						return nil, err
-					}
-					if isNew := c.markSeen(m); !isNew {
-						continue
+		if err != nil {
+			return err
+		}
+
+		// Holds column to its value mapping for a row and the entire thing will be
+		// JSON encoded later.
+		value := make(map[string]interface{})
+
+		// Holds the list of values of primary keys.
+		key := make([]interface{}, 0)
+
+		for k, v := range row {
+
+			switch vv := v.(type) {
```

Comment:
There is too much info/and nesting with error handling inside this switch.
Let's pull this entire switch into helper function; something like
... `decodeParquetValueAsJson(interface...)`